#this is a script for generating RNA velocity estimations using velocyto.R
#this tutorial follows the vignette for the veolcyto.R wrapper in seurat(link: https://htmlpreview.github.io/?https://github.com/satijalab/seurat.wrappers/blob/master/docs/velocity.html)

#load relavant libraries and datasets
#for libraries, we will need to install the seurat wrapper for veolcyto.R 
#for datasets we will need the loom matricies generated by the velocyto.Py commmand line interface
library(Seurat)
library(devtools)
library(velocyto.R)
library(SeuratWrappers)

#load BAM derived loom matrices
trachea.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/trachea_possorted_genome_bam_VDY13.loom")
old.3dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_3dpi_possorted_genome_bam_NKMBN.loom")
old.5dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_5dpi_possorted_genome_bam_FSK6X.loom")
old.7dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_7dpi_possorted_genome_bam_OJCZL.loom")
old.9dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_9dpi_possorted_genome_bam_OBAL3.loom")
old.11dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_11dpi_possorted_genome_bam_ZBDY1.loom")
old.14dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_14dpi_possorted_genome_bam_ADK95.loom")
old.17dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_17dpi_possorted_genome_bam_WM7N3.loom")
old.21dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_21dpi_possorted_genome_bam_9MF09.loom")
old.nepi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/old_nepi_possorted_genome_bam_MPV2T.loom")
redo.4mpi.T.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_4mpi_possorted_genome_bam_1Y3JR.loom")
redo.8mpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_8mpi_possorted_genome_bam_9U1WW.loom")
redo.17dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_17dpi_possorted_genome_bam_9543Q.loom")
redo.21dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_21dpi_possorted_genome_bam_YS0SH.loom")
redo.n_3_5_7dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_91114dpi_possorted_genome_bam_UL7UH.loom")
redo.9_11_14dpi.loom <- ReadVelocity(file = "/home/stripplab2/Desktop/20190910.secondaryanalysis/20190910.velocyto/loomfile/redo_n357dpi_possorted_genome_bam_U42TY.loom")

#before converting the loom files into Seurat objects we will need to perform some dataframe manipulation on the cell barcodes
#this is to ensure that RNA velocities calculated from the loom matrices can to matched to its corresponding cell on the filtered gene mextrix derived daatset
#remove the prefix/suffix identifiers in front of each cell barcode in the loom matrix
head(colnames(trachea.loom$spliced))
head(colnames(trachea.loom$unspliced))
head(colnames(trachea.loom$ambiguous))
colnames(trachea.loom$spliced) <- paste(substring(colnames(trachea.loom$spliced),28,43),sep="")
colnames(trachea.loom$unspliced) <- paste(substring(colnames(trachea.loom$unspliced),28,43),sep="")
colnames(trachea.loom$ambiguous) <- paste(substring(colnames(trachea.loom$ambiguous),28,43),sep="")

head(colnames(redo.n_3_5_7dpi.loom$spliced))
head(colnames(redo.n_3_5_7dpi.loom$unspliced))
head(colnames(redo.n_3_5_7dpi.loom$ambiguous))
colnames(redo.n_3_5_7dpi.loom$spliced) <- paste(substring(colnames(redo.n_3_5_7dpi.loom$spliced),28,43),sep="")
colnames(redo.n_3_5_7dpi.loom$unspliced) <- paste(substring(colnames(redo.n_3_5_7dpi.loom$unspliced),28,43),sep="")
colnames(redo.n_3_5_7dpi.loom$ambiguous) <- paste(substring(colnames(redo.n_3_5_7dpi.loom$ambiguous),28,43),sep="")

head(colnames(redo.9_11_14dpi.loom$spliced))
head(colnames(redo.9_11_14dpi.loom$unspliced))
head(colnames(redo.9_11_14dpi.loom$ambiguous))
colnames(redo.9_11_14dpi.loom$spliced) <- paste(substring(colnames(redo.9_11_14dpi.loom$spliced),28,43),sep="")
colnames(redo.9_11_14dpi.loom$unspliced) <- paste(substring(colnames(redo.9_11_14dpi.loom$unspliced),28,43),sep="")
colnames(redo.9_11_14dpi.loom$ambiguous) <- paste(substring(colnames(redo.9_11_14dpi.loom$ambiguous),28,43),sep="")

head(colnames(redo.8mpi.loom$spliced))
head(colnames(redo.8mpi.loom$unspliced))
head(colnames(redo.8mpi.loom$ambiguous))
colnames(redo.8mpi.loom$spliced) <- paste(substring(colnames(redo.8mpi.loom$spliced),28,43),sep="")
colnames(redo.8mpi.loom$unspliced) <- paste(substring(colnames(redo.8mpi.loom$unspliced),28,43),sep="")
colnames(redo.8mpi.loom$ambiguous) <- paste(substring(colnames(redo.8mpi.loom$ambiguous),28,43),sep="")

head(colnames(redo.4mpi.T.loom$spliced))
head(colnames(redo.4mpi.T.loom$unspliced))
head(colnames(redo.4mpi.T.loom$ambiguous))
colnames(redo.4mpi.T.loom$spliced) <- paste(substring(colnames(redo.4mpi.T.loom$spliced),28,43),sep="")
colnames(redo.4mpi.T.loom$unspliced) <- paste(substring(colnames(redo.4mpi.T.loom$unspliced),28,43),sep="")
colnames(redo.4mpi.T.loom$ambiguous) <- paste(substring(colnames(redo.4mpi.T.loom$ambiguous),28,43),sep="")

head(colnames(redo.21dpi.loom$spliced))
head(colnames(redo.21dpi.loom$unspliced))
head(colnames(redo.21dpi.loom$ambiguous))
colnames(redo.21dpi.loom$spliced) <- paste(substring(colnames(redo.21dpi.loom$spliced),28,43),sep="")
colnames(redo.21dpi.loom$unspliced) <- paste(substring(colnames(redo.21dpi.loom$unspliced),28,43),sep="")
colnames(redo.21dpi.loom$ambiguous) <- paste(substring(colnames(redo.21dpi.loom$ambiguous),28,43),sep="")

head(colnames(redo.17dpi.loom$spliced))
head(colnames(redo.17dpi.loom$unspliced))
head(colnames(redo.17dpi.loom$ambiguous))
colnames(redo.17dpi.loom$spliced) <- paste(substring(colnames(redo.17dpi.loom$spliced),28,43),sep="")
colnames(redo.17dpi.loom$unspliced) <- paste(substring(colnames(redo.17dpi.loom$unspliced),28,43),sep="")
colnames(redo.17dpi.loom$ambiguous) <- paste(substring(colnames(redo.17dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.nepi.loom$spliced))
head(colnames(old.nepi.loom$unspliced))
head(colnames(old.nepi.loom$ambiguous))
colnames(old.nepi.loom$spliced) <- paste(substring(colnames(old.nepi.loom$spliced),28,43),sep="")
colnames(old.nepi.loom$unspliced) <- paste(substring(colnames(old.nepi.loom$unspliced),28,43),sep="")
colnames(old.nepi.loom$ambiguous) <- paste(substring(colnames(old.nepi.loom$ambiguous),28,43),sep="")

head(colnames(old.9dpi.loom$spliced))
head(colnames(old.9dpi.loom$unspliced))
head(colnames(old.9dpi.loom$ambiguous))
colnames(old.9dpi.loom$spliced) <- paste(substring(colnames(old.9dpi.loom$spliced),28,43),sep="")
colnames(old.9dpi.loom$unspliced) <- paste(substring(colnames(old.9dpi.loom$unspliced),28,43),sep="")
colnames(old.9dpi.loom$ambiguous) <- paste(substring(colnames(old.9dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.7dpi.loom$spliced))
head(colnames(old.7dpi.loom$unspliced))
head(colnames(old.7dpi.loom$ambiguous))
colnames(old.7dpi.loom$spliced) <- paste(substring(colnames(old.7dpi.loom$spliced),28,43),sep="")
colnames(old.7dpi.loom$unspliced) <- paste(substring(colnames(old.7dpi.loom$unspliced),28,43),sep="")
colnames(old.7dpi.loom$ambiguous) <- paste(substring(colnames(old.7dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.5dpi.loom$spliced))
head(colnames(old.5dpi.loom$unspliced))
head(colnames(old.5dpi.loom$ambiguous))
colnames(old.5dpi.loom$spliced) <- paste(substring(colnames(old.5dpi.loom$spliced),28,43),sep="")
colnames(old.5dpi.loom$unspliced) <- paste(substring(colnames(old.5dpi.loom$unspliced),28,43),sep="")
colnames(old.5dpi.loom$ambiguous) <- paste(substring(colnames(old.5dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.3dpi.loom$spliced))
head(colnames(old.3dpi.loom$unspliced))
head(colnames(old.3dpi.loom$ambiguous))
colnames(old.3dpi.loom$spliced) <- paste(substring(colnames(old.3dpi.loom$spliced),28,43),sep="")
colnames(old.3dpi.loom$unspliced) <- paste(substring(colnames(old.3dpi.loom$unspliced),28,43),sep="")
colnames(old.3dpi.loom$ambiguous) <- paste(substring(colnames(old.3dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.21dpi.loom$spliced))
head(colnames(old.21dpi.loom$unspliced))
head(colnames(old.21dpi.loom$ambiguous))
colnames(old.21dpi.loom$spliced) <- paste(substring(colnames(old.21dpi.loom$spliced),28,43),sep="")
colnames(old.21dpi.loom$unspliced) <- paste(substring(colnames(old.21dpi.loom$unspliced),28,43),sep="")
colnames(old.21dpi.loom$ambiguous) <- paste(substring(colnames(old.21dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.17dpi.loom$spliced))
head(colnames(old.17dpi.loom$unspliced))
head(colnames(old.17dpi.loom$ambiguous))
colnames(old.17dpi.loom$spliced) <- paste(substring(colnames(old.17dpi.loom$spliced),28,43),sep="")
colnames(old.17dpi.loom$unspliced) <- paste(substring(colnames(old.17dpi.loom$unspliced),28,43),sep="")
colnames(old.17dpi.loom$ambiguous) <- paste(substring(colnames(old.17dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.14dpi.loom$spliced))
head(colnames(old.14dpi.loom$unspliced))
head(colnames(old.14dpi.loom$ambiguous))
colnames(old.14dpi.loom$spliced) <- paste(substring(colnames(old.14dpi.loom$spliced),28,43),sep="")
colnames(old.14dpi.loom$unspliced) <- paste(substring(colnames(old.14dpi.loom$unspliced),28,43),sep="")
colnames(old.14dpi.loom$ambiguous) <- paste(substring(colnames(old.14dpi.loom$ambiguous),28,43),sep="")

head(colnames(old.11dpi.loom$spliced))
head(colnames(old.11dpi.loom$unspliced))
head(colnames(old.11dpi.loom$ambiguous))
colnames(old.11dpi.loom$spliced) <- paste(substring(colnames(old.11dpi.loom$spliced),28,43),sep="")
colnames(old.11dpi.loom$unspliced) <- paste(substring(colnames(old.11dpi.loom$unspliced),28,43),sep="")
colnames(old.11dpi.loom$ambiguous) <- paste(substring(colnames(old.11dpi.loom$ambiguous),28,43),sep="")

#add a suffix identifier to make the cell barcodes match up to the cell barcodes from the filtered gene matrix derived dataset
head(colnames(trachea.loom$spliced))
head(colnames(trachea.loom$unspliced))
head(colnames(trachea.loom$ambiguous))
colnames(trachea.loom$spliced) <- paste0(colnames(trachea.loom$spliced),"_7")
colnames(trachea.loom$unspliced) <- paste0(colnames(trachea.loom$unspliced),"_7")
colnames(trachea.loom$ambiguous) <- paste0(colnames(trachea.loom$ambiguous),"_7")

head(colnames(redo.n_3_5_7dpi.loom$spliced))
head(colnames(redo.n_3_5_7dpi.loom$unspliced))
head(colnames(redo.n_3_5_7dpi.loom$ambiguous))
colnames(redo.n_3_5_7dpi.loom$spliced) <- paste0(colnames(redo.n_3_5_7dpi.loom$spliced),"_1")
colnames(redo.n_3_5_7dpi.loom$unspliced) <- paste0(colnames(redo.n_3_5_7dpi.loom$unspliced),"_1")
colnames(redo.n_3_5_7dpi.loom$ambiguous) <- paste0(colnames(redo.n_3_5_7dpi.loom$ambiguous),"_1")

head(colnames(redo.9_11_14dpi.loom$spliced))
head(colnames(redo.9_11_14dpi.loom$unspliced))
head(colnames(redo.9_11_14dpi.loom$ambiguous))
colnames(redo.9_11_14dpi.loom$spliced) <- paste0(colnames(redo.9_11_14dpi.loom$spliced),"_2")
colnames(redo.9_11_14dpi.loom$unspliced) <- paste0(colnames(redo.9_11_14dpi.loom$unspliced),"_2")
colnames(redo.9_11_14dpi.loom$ambiguous) <- paste0(colnames(redo.9_11_14dpi.loom$ambiguous),"_2")

head(colnames(redo.8mpi.loom$spliced))
head(colnames(redo.8mpi.loom$unspliced))
head(colnames(redo.8mpi.loom$ambiguous))
colnames(redo.8mpi.loom$spliced) <- paste0(colnames(redo.8mpi.loom$spliced),"_6")
colnames(redo.8mpi.loom$unspliced) <- paste0(colnames(redo.8mpi.loom$unspliced),"_6")
colnames(redo.8mpi.loom$ambiguous) <- paste0(colnames(redo.8mpi.loom$ambiguous),"_6")

head(colnames(redo.4mpi.T.loom$spliced))
head(colnames(redo.4mpi.T.loom$unspliced))
head(colnames(redo.4mpi.T.loom$ambiguous))
colnames(redo.4mpi.T.loom$spliced) <- paste0(colnames(redo.4mpi.T.loom$spliced),"_3")
colnames(redo.4mpi.T.loom$unspliced) <- paste0(colnames(redo.4mpi.T.loom$unspliced),"_3")
colnames(redo.4mpi.T.loom$ambiguous) <- paste0(colnames(redo.4mpi.T.loom$ambiguous),"_3")

head(colnames(redo.21dpi.loom$spliced))
head(colnames(redo.21dpi.loom$unspliced))
head(colnames(redo.21dpi.loom$ambiguous))
colnames(redo.21dpi.loom$spliced) <- paste0(colnames(redo.21dpi.loom$spliced),"_5")
colnames(redo.21dpi.loom$unspliced) <- paste0(colnames(redo.21dpi.loom$unspliced),"_5")
colnames(redo.21dpi.loom$ambiguous) <- paste0(colnames(redo.21dpi.loom$ambiguous),"_5")

head(colnames(redo.17dpi.loom$spliced))
head(colnames(redo.17dpi.loom$unspliced))
head(colnames(redo.17dpi.loom$ambiguous))
colnames(redo.17dpi.loom$spliced) <- paste0(colnames(redo.17dpi.loom$spliced),"_4")
colnames(redo.17dpi.loom$unspliced) <- paste0(colnames(redo.17dpi.loom$unspliced),"_4")
colnames(redo.17dpi.loom$ambiguous) <- paste0(colnames(redo.17dpi.loom$ambiguous),"_4")

head(colnames(old.nepi.loom$spliced))
head(colnames(old.nepi.loom$unspliced))
head(colnames(old.nepi.loom$ambiguous))
colnames(old.nepi.loom$spliced) <- paste0(colnames(old.nepi.loom$spliced),"_8")
colnames(old.nepi.loom$unspliced) <- paste0(colnames(old.nepi.loom$unspliced),"_8")
colnames(old.nepi.loom$ambiguous) <- paste0(colnames(old.nepi.loom$ambiguous),"_8")

head(colnames(old.9dpi.loom$spliced))
head(colnames(old.9dpi.loom$unspliced))
head(colnames(old.9dpi.loom$ambiguous))
colnames(old.9dpi.loom$spliced) <- paste0(colnames(old.9dpi.loom$spliced),"_12")
colnames(old.9dpi.loom$unspliced) <- paste0(colnames(old.9dpi.loom$unspliced),"_12")
colnames(old.9dpi.loom$ambiguous) <- paste0(colnames(old.9dpi.loom$ambiguous),"_12")

head(colnames(old.7dpi.loom$spliced))
head(colnames(old.7dpi.loom$unspliced))
head(colnames(old.7dpi.loom$ambiguous))
colnames(old.7dpi.loom$spliced) <- paste0(colnames(old.7dpi.loom$spliced),"_11")
colnames(old.7dpi.loom$unspliced) <- paste0(colnames(old.7dpi.loom$unspliced),"_11")
colnames(old.7dpi.loom$ambiguous) <- paste0(colnames(old.7dpi.loom$ambiguous),"_11")

head(colnames(old.5dpi.loom$spliced))
head(colnames(old.5dpi.loom$unspliced))
head(colnames(old.5dpi.loom$ambiguous))
colnames(old.5dpi.loom$spliced) <- paste0(colnames(old.5dpi.loom$spliced),"_10")
colnames(old.5dpi.loom$unspliced) <- paste0(colnames(old.5dpi.loom$unspliced),"_10")
colnames(old.5dpi.loom$ambiguous) <- paste0(colnames(old.5dpi.loom$ambiguous),"_10")

head(colnames(old.3dpi.loom$spliced))
head(colnames(old.3dpi.loom$unspliced))
head(colnames(old.3dpi.loom$ambiguous))
colnames(old.3dpi.loom$spliced) <- paste0(colnames(old.3dpi.loom$spliced),"_9")
colnames(old.3dpi.loom$unspliced) <- paste0(colnames(old.3dpi.loom$unspliced),"_9")
colnames(old.3dpi.loom$ambiguous) <- paste0(colnames(old.3dpi.loom$ambiguous),"_9")

head(colnames(old.21dpi.loom$spliced))
head(colnames(old.21dpi.loom$unspliced))
head(colnames(old.21dpi.loom$ambiguous))
colnames(old.21dpi.loom$spliced) <- paste0(colnames(old.21dpi.loom$spliced),"_16")
colnames(old.21dpi.loom$unspliced) <- paste0(colnames(old.21dpi.loom$unspliced),"_16")
colnames(old.21dpi.loom$ambiguous) <- paste0(colnames(old.21dpi.loom$ambiguous),"_16")

head(colnames(old.17dpi.loom$spliced))
head(colnames(old.17dpi.loom$unspliced))
head(colnames(old.17dpi.loom$ambiguous))
colnames(old.17dpi.loom$spliced) <- paste0(colnames(old.17dpi.loom$spliced),"_15")
colnames(old.17dpi.loom$unspliced) <- paste0(colnames(old.17dpi.loom$unspliced),"_15")
colnames(old.17dpi.loom$ambiguous) <- paste0(colnames(old.17dpi.loom$ambiguous),"_15")

head(colnames(old.14dpi.loom$spliced))
head(colnames(old.14dpi.loom$unspliced))
head(colnames(old.14dpi.loom$ambiguous))
colnames(old.14dpi.loom$spliced) <- paste0(colnames(old.14dpi.loom$spliced),"_14")
colnames(old.14dpi.loom$unspliced) <- paste0(colnames(old.14dpi.loom$unspliced),"_14")
colnames(old.14dpi.loom$ambiguous) <- paste0(colnames(old.14dpi.loom$ambiguous),"_14")

head(colnames(old.11dpi.loom$spliced))
head(colnames(old.11dpi.loom$unspliced))
head(colnames(old.11dpi.loom$ambiguous))
colnames(old.11dpi.loom$spliced) <- paste0(colnames(old.11dpi.loom$spliced),"_13")
colnames(old.11dpi.loom$unspliced) <- paste0(colnames(old.11dpi.loom$unspliced),"_13")
colnames(old.11dpi.loom$ambiguous) <- paste0(colnames(old.11dpi.loom$ambiguous),"_13")

#the next step is to convert the loom matrices into seurat objects
#make sure you so this before attempting to merge the datasets. otherwise, the merged dataset will use up too much of the computers memory and abort the R session
trachea.loom <- as.Seurat(x = trachea.loom)
old.3dpi.loom <- as.Seurat(x = old.3dpi.loom)
old.5dpi.loom <- as.Seurat(x = old.5dpi.loom)
old.7dpi.loom <- as.Seurat(x = old.7dpi.loom)
old.9dpi.loom <- as.Seurat(x = old.9dpi.loom)
old.11dpi.loom <- as.Seurat(x = old.11dpi.loom)
old.14dpi.loom <- as.Seurat(x = old.14dpi.loom)
old.17dpi.loom <- as.Seurat(x = old.17dpi.loom)
old.21dpi.loom <- as.Seurat(x = old.21dpi.loom)
old.nepi.loom <- as.Seurat(x = old.nepi.loom)
redo.4mpi.T.loom <- as.Seurat(x = redo.4mpi.T.loom)
redo.8mpi.loom <- as.Seurat(x = redo.8mpi.loom)
redo.17dpi.loom <- as.Seurat(x = redo.17dpi.loom)
redo.21dpi.loom <- as.Seurat(x = redo.21dpi.loom)
redo.n_3_5_7dpi.loom <- as.Seurat(x = redo.n_3_5_7dpi.loom)
redo.9_11_14dpi.loom <- as.Seurat(x = redo.9_11_14dpi.loom)

#merge seurat objects
epitimecourse.preintegration <- merge(x = trachea.loom, y = c(old.nepi.loom, old.3dpi.loom, old.5dpi.loom, old.7dpi.loom, old.9dpi.loom, old.11dpi.loom, old.14dpi.loom, old.21dpi.loom, redo.n_3_5_7dpi.loom, redo.9_11_14dpi.loom, redo.4mpi.T.loom, redo.8mpi.loom, redo.17dpi.loom, redo.21dpi.loom))

#perform standard seurat workflow(ie. data nomalization and clustering)
#dataset integration (to correct for batch effect) is not needed 
#SCTransform is preferred to standard log normalization
#keep in mind that SCTransform function takes the spliced matrix as an input but the downstream anlaysis draws from BOTH the spliced and unspliced matrices to calculate RNA velocity
#this means if you filter any cells before calculating RNA velocity, you have to filter them from both spliced and unspliced matrices. This step is essential to prevent generating NAs values in the final matrix.
#Typically, its safer to just analyze everything since the velocities are going to be projected onto a dataset where standard QCs have been applied
epitimecourse.preintegration <- RunPCA(object = epitimecourse.preintegration, verbose = FALSE)
epitimecourse.preintegration <- FindNeighbors(object = epitimecourse.preintegration, dims = 1:20)
epitimecourse.preintegration <- FindClusters(object = epitimecourse.preintegration)
epitimecourse.preintegration <- RunUMAP(object = epitimecourse.preintegration, dims = 1:20)
DimPlot(epitimecourse.preintegration)

saveRDS(epitimecourse.preintegration, file ="epitimecourse.velocyto.V5.rds")

#Cell trajectory analysis will be performed using scVelo 
#http://htmlpreview.github.io/?https://github.com/satijalab/seurat-wrappers/blob/master/docs/scvelo.html
#load in dataset and libraries
library(Seurat)
library(SeuratDisk)
library(SeuratWrappers)

velo <- readRDS("C:/Users/Stripp Lab 2/Desktop/20220815_scVelo/epitimecourse.velocyto.V5.rds")
influenza_timecourse_epithelial <- readRDS("/Volumes/20190816.singlecellRNAseq/20220801_Geo_upload_V2/influenza_timecourse_epithelial.rds")
influenza_timecourse_epithelial@meta.data

#match cell barcodes from "velo" to "influenza_timecourse_epithelial"
#This step is needed to transfer metadata from "influenza_timecourse_epithelial" to "velo"
#generate a seperate dataset with metadata information and corresponding cell barcode.
subset2 <- subset(influenza_timecourse_epithelial, idents=c("Basal", "Serous","Club"))
meta <- subset2@meta.data
anno <- meta[ ,"anno", drop=F]
timepoint <- meta[ ,"timepoint", drop=F]

#concatenate meta data information from 'influenza_timecourse_epithelial' onto 'velo'
velo$anno <- anno
velo$timepoint <- timepoint
velo@meta.data

table(influenza_timecourse_epithelial@meta.data$timepoint)

#subset conditions of interest and recluster
Idents(velo) <- "timepoint"
early <- subset(velo, idents = c("trachea", "3 dpi", "5 dpi", "7 dpi", "9 dpi", "11 dpi", "14 dpi"))
late <- subset(velo, idents = c("14 dpi", "17 dpi", "21 dpi", "14 dpi"))
early <- RunUMAP(early, dims = 1:20)
late <- RunUMAP(late, dims = 1:20)
DimPlot(early, group.by="anno")
DimPlot(late, group.by="anno")

#save
SaveH5Seurat(early, filename = "Early-BC-Serous-Club.h5Seurat")
SaveH5Seurat(late, filename = "Early-BC-Serous-Club.h5Seurat")
Convert("Early-BC-Serous-Club.h5Seurat", dest = "h5ad")
Convert("Late-BC-Serous-Club.h5Seurat", dest = "h5ad")

# In Python
import scvelo as scv
adata = scv.read("Early-BC-Serous-Club.h5Seurat")
adata
scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)
scv.pp.moments(adata, n_pcs=30, n_neighbors=30)
scv.tl.velocity(adata)
scv.tl.velocity_graph(adata)
scv.pl.velocity_embedding_stream(adata, basis="umap", color="seurat_clusters")